source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst

```diff
@@ -0,0 +1,199 @@
+      private Pane root;
+
+      @FXML
+      private Slider xSlider;
+
+      @FXML
+      private Slider ySlider;
+   }
+
+In order to display our pane on our custom widget we need to override the ``getView()`` method and return our ``StackedPane``.
+
+.. code-block:: java
+
+   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
+   import javafx.fxml.FXML;
+   import edu.wpi.first.shuffleboard.api.widget.Description;
+   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
+
+   @Description(name = "MyPoint2D", dataTypes = MyPoint2D.class)
+   @ParamatrizedController("Point2DWidget.fxml")
+   public final class Point2DWidget extends SimpleAnnotatedWidget<MyPoint2D> {
+
+      @FXML
+      private StackPane root;
+
+      @FXML
+      private Slider xSlider;
+
+      @FXML
+      private Slider ySlider;
+
+      @Override
+      public Pane getView() {
+         return root;
+      }
+
+   }
+
+Binding Elements and Adding Listeners
+-------------------------------------
+Binding is a mechanism that allows JavaFX widgets to express direct relationships with the data source. For example, changing a widget will change its related NetworkTableEntry and vise versa.
+
+An example, in this case, would be changing the X and Y coordinate of our 2D point by changing out xSlider and ySlider respectively.
+
+A good practice is to set bindings in the ``initalize()`` method tagged with the ``@FXML`` annotation which will be called by the FXML framework.
+
+.. code-block:: java
+
+   import edu.wpi.first.shuffleboard.api.widget.SimpleAnnotatedWidget;
+   import javafx.fxml.FXML;
+   import edu.wpi.first.shuffleboard.api.widget.Description;
+   import edu.wpi.first.shuffleboard.api.widget.ParametrizedController;
+
+   @Description(name = "MyPoint2D", dataTypes = MyPoint2D.class)
+   @ParamatrizedController("Point2DWidget.fxml")
+   public final class Point2DWidget extends SimpleAnnotatedWidget<MyPoint2D> {
+
+      @FXML
+      private StackPane root;
+
+      @FXML
+      private Slider xSlider;
+
+      @FXML
+      private Slider ySlider;
+
+      @FXML
+      private void initialize() {
+         xSlider.valueProperty().bind(dataOrDefault.map(MyPoint2D::getX));
+         ySlider.valueProperty().bind(dataOrDefault.map(MyPoint2D::getY));
+      }
+
+      @Override
+      public Pane getView() {
+         return root;
+      }
+
+    }
+
+The above ``initalize`` method binds the slider's value property to the ``MyPoint2D`` data class' corresponding X and Y value. Meaning, changing the slider will change the coordinate and vise versa.
+The ``dataOrDefault.map()`` method will get the data source's value, or, if no source is present, will return the default value.
+
+Using a listener is another way to change values when the slider or data source has changed. For example a listener for our slider would be:
+
+.. code-block:: java
+
+   xSlider.valueProperty().addListener((observable, oldValue, newValue) -> setData(newValue));
```

Comment:
Since the data type is `Point2D`, `setData(double)` won't work

```suggestion
   xSlider.valueProperty().addListener((observable, oldValue, newValue) -> setData(getData().withX(newValue));
```